biscuit-parser/src/builder.rs

```diff
@@ -328,33 +328,38 @@ impl Rule {
     }
 
     pub fn validate_variables(&self) -> Result<(), String> {
```

Comment:
this would need to return a 3-valued enum (Head, Body, Both) to signal where the free variables were found, in order to only return the relevant span in the parsing error.